python/rmm/_lib/memory_resource.pyx

```diff
@@ -386,35 +430,42 @@ cpdef get_per_device_resource(int device):
     return _per_device_mrs[device]
 
 
-cpdef _set_per_device_resource(int device, MemoryResource mr):
+cpdef set_per_device_resource(int device, DeviceMemoryResource mr):
     """
     Set the default memory resource for the specified device.
 
     Parameters
     ----------
     device : int
         The ID of the device for which to get the memory resource.
-    mr : MemoryResource
+    mr : DeviceMemoryResource
         The memory resource to set.  Must have been created while device was
         the active CUDA device.
     """
     global _per_device_mrs
     _per_device_mrs[device] = mr
-    _mr = mr  # coerce Python object to C object
-    set_per_device_resource(device, _mr.c_obj)
+
+    # Since cuda_device_id does not have a default constructor, it must be heap
+    # allocated
+    cdef cuda_device_id* device_id
+    try:
+        device_id = new cuda_device_id(device)
+        _set_per_device_resource(deref(device_id), mr.get_mr())
+    finally:
+        del device_id
```

Comment:
Could we use a smart pointer here instead of `new` / `del`?

Reply:
Fixed